### DISCUSSION TITLE
Method not allowed on read_csv

### DISCUSSION THREAD

**Discussion by Lukacms** [duckdb/duckdb#10903] [Q&A]
### What happens?

I want to read the content of an api endpoint with `read_csv` in a sql file using duckdb. When I try to do that,  I have the following error:
```sql
D select * from read_csv("endpoint");
Error: IO Error: Unable to connect to URL "endpoint": 405 (Method Not Allowed)
```

The headers of the endpoint I'm trying to get are the following:
```sh
* IPv6: (none)
* IPv4
*   Trying 
* Connected to `api host` () port 80
> GET /endpoint HTTP/1.1
> Host: `api host`
> User-Agent: curl/8.6.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: text/csv
< Date: Wed, 28 Feb 2024 16:57:09 GMT
< Transfer-Encoding: chunked
< 
{ [1340 bytes data]
100 40.7M    0 40.7M    0     0  8056k      0 --:--:--  0:00:05 --:--:--  9.8M* Leftovers after chunking: 15 bytes
100 52.2M    0 52.2M    0     0  8818k      0 --:--:--  0:00:06 --:--:-- 11.0M
* Connection #0 to host left intact
```

It's a golang api, made with [`httprouter`](https://github.com/julienschmidt/httprouter)
I guess the problem is on our side, I just would like to know if I'm supposed to add / modify headers in order to be able to use `read_csv` ? Or just what we are doing wrong :sweat_smile: 

Note that I am able to fetch the data in python with `pandas.read_csv(url)`, I'd just like to be able to do it on duckdb

[This issue](https://github.com/duckdb/duckdb/issues/3991) may be linked, but I tried exploring that, without success

### To Reproduce

I am not able to give the specific url, as it's a company private api :/

The closest would be with a csv sample website, such as https://filesamples.com/samples/document/csv/sample1.csv. The header returned by curl is:
```
$> curl -v "https://filesamples.com/samples/document/csv/sample1.csv"
> GET /samples/document/csv/sample1.csv HTTP/2
> Host: filesamples.com
> User-Agent: curl/8.6.0
> Accept: */*
> 
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [230 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [230 bytes data]
* old SSL session ID is stale, removing
{ [5 bytes data]
< HTTP/2 200 
< date: Wed, 28 Feb 2024 16:51:05 GMT
< content-type: application/octet-stream
< content-length: 502
< cache-control: public, max-age=2592000
< display: staticcontent_sol
< etag: "64a5bc02-1f6-gzip"
< last-modified: Wed, 05 Jul 2023 18:52:50 GMT
< response: 200
< vary: Accept-Encoding,Origin
< x-middleton-display: staticcontent_sol
< x-middleton-response: 200
< cf-cache-status: HIT
< age: 200395
< accept-ranges: bytes
< report-to: 
< nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< server: cloudflare
< cf-ray: 85ca2b9788966f7c-CDG
< alt-svc: h3=":443"; ma=86400
```

And read_csv works here, so the problem is on our side. Is there any specific header to put ? Adding `Accept-Range` to the headers did not change anything

### OS:

linux-arch64

### DuckDB Version:

v0.10.0 20b1486d11

### DuckDB Client:

duckdb cli

### Full Name:

Luka Camus

### Affiliation:

Displayce

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a release build (and could not test with a nightly build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by szarnyasg** [comment]
Hi @Lukacms, I moved this to the Q&A section. We'll take a look – and this may need documentation – but this is likely not a bug.

**Comment by Tishj** [comment]
My quick guess is that the preflight range request we send is getting blocked

**Comment by Lukacms** [comment]
Thank you for the answers ! We'll test that as soon as possible :)

**Comment by Lukacms** [comment]
Hello, we solved the problem, which was on our side. I'll just write what we did, if anyone has the same issue.

The problem was because our endpoint didn't handle range requests, which conflicted with duckdb's `read_csv` request, so instead of just calling:
```go
w.Write([]byte(body))
```
with `w` being [http.ResponseWriter](https://pkg.go.dev/io), we called the following method of [httprouter](https://pkg.go.dev/github.com/julienschmidt/httprouter)
```go
http.ServeContent(w, r, "bid-requests-metrics.csv", time.Now(), bytes.NewReader([]byte(body)))
```
with w being the `ResponseWriter`, r the `Request` and `body` being a string.

Thank you again for your hints :)
